plugins/statistics.koplugin/main.lua

```diff
@@ -73,58 +83,350 @@ function ReaderStatistics:init()
+                    text =_("New version of statistics plugin is detected.\n" ..
+                        "A new database has been created.")})
+            end
+        end
+        self:saveSettings()
     end
-    return props
 end
 
-function ReaderStatistics:initData(config)
+function ReaderStatistics:initData()
     if self:isDocless() or not self.is_enabled then
         return
     end
     -- first execution
     if not self.data then
         self.data = { performance_in_pages= {} }
-        self:inplaceMigration();  -- first time merge data
     end
-
     local book_properties = self:getBookProperties()
     self.data.title = book_properties.title
+    if self.data.title == nil or self.data.title == "" then
+        self.data.title = self.document.file:match("^.+/(.+)$")
+    end
     self.data.authors = book_properties.authors
     self.data.language = book_properties.language
     self.data.series = book_properties.series
 
     self.data.pages = self.view.document:getPageCount()
-    return
+    self.data.md5 = self:md5(self.document.file)
+    self.curr_total_time = 0
+    self.curr_total_pages = 0
+    self.id_curr_book = self:getIdBookDB()
+    self.total_read_pages, self.total_read_time = self:getPageTimeTotalStats(self.id_curr_book)
+    if self.total_read_pages > 0 then
+        local avg_time = math.ceil(self.total_read_time / self.total_read_pages)
+        self.avg_time = avg_time
+        self.ui:handleEvent(Event:new("UpdateStats", avg_time))
+        self.view.footer:updateFooter()
+    end
 end
 
-local function generateReadBooksTable(title, dates)
-    local result = {}
-    for k, v in tableutil.spairs(dates, function(t, a, b) return t[b].date < t[a].date end) do
-        table.insert(result, {
-            k,
-            T(_("Pages (%1) Time: %2"), v.count, util.secondsToClock(v.read, false))
-        })
+function ReaderStatistics:md5(file)
+    if file == nil then
+        return nil
+    end
+    local md5 = require("ffi/MD5")
+    local lshift = bit.lshift
+    local step, size = 1024, 1024
+    local m = md5.new()
+    local file_handle = io.open(file, 'rb')
+    for i = -1, 10 do
+        file_handle:seek("set", lshift(step, 2*i))
+        local sample = file_handle:read(size)
+        if sample then
+            m:update(sample)
+        else
+            break
+        end
+    end
+    return m:sum()
+end
+
+function ReaderStatistics:createDB()
+    local conn = SQ3.open(db_location)
+    conn:exec("CREATE TABLE IF NOT EXISTS book(id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT, authors TEXT," ..
+        "notes INTEGER, last_open INTEGER, highlights INTEGER, pages INTEGER, series TEXT, language TEXT," ..
+        "md5 TEXT, total_read_time INTEGER, total_read_pages INTEGER);" ..
+        "CREATE TABLE IF NOT EXISTS page_stat(id_book INTEGER, page INTEGER NOT NULL, start_time INTEGER NOT NULL, " ..
+        "period INTEGER NOT NULL, UNIQUE (page, start_time), FOREIGN KEY(id_book) REFERENCES book(id));" ..
```

Comment:
I create period because it's much easier and faster to make SQL statements for statistic like "Today Period", "Current Period", Time in Months, Weeks etc.
For example - sum(period):

```
SELECT sum(period), count(DISTINCT page) 
FROM page_stat 
WHERE id_book = 'id'
```